optuna/samplers/nsgaiii.py

```diff
@@ -0,0 +1,543 @@
+                additional_elite_population = _niching(
+                    target_population_size,
+                    population,
+                    cnt2refs,
+                    ref2pops,
+                )
+                elite_population.extend(additional_elite_population)
+                break
+        return elite_population
+
+    def _fast_non_dominated_sort(
+        self,
+        population: List[FrozenTrial],
+        directions: List[optuna.study.StudyDirection],
+    ) -> List[List[FrozenTrial]]:
+        if self._constraints_func is not None:
+            for _trial in population:
+                _constraints = _trial.system_attrs.get(_CONSTRAINTS_KEY)
+                if _constraints is None:
+                    continue
+                if np.any(np.isnan(np.array(_constraints))):
+                    raise ValueError("NaN is not acceptable as constraint value.")
+
+        dominated_count: DefaultDict[int, int] = defaultdict(int)
+        dominates_list = defaultdict(list)
+
+        dominates = _dominates if self._constraints_func is None else _constrained_dominates
+
+        for p, q in itertools.combinations(population, 2):
+            if dominates(p, q, directions):
+                dominates_list[p.number].append(q.number)
+                dominated_count[q.number] += 1
+            elif dominates(q, p, directions):
+                dominates_list[q.number].append(p.number)
+                dominated_count[p.number] += 1
+
+        population_per_rank = []
+        while population:
+            non_dominated_population = []
+            i = 0
+            while i < len(population):
+                if dominated_count[population[i].number] == 0:
+                    individual = population[i]
+                    if i == len(population) - 1:
+                        population.pop()
+                    else:
+                        population[i] = population.pop()
+                    non_dominated_population.append(individual)
+                else:
+                    i += 1
+
+            for x in non_dominated_population:
+                for y in dominates_list[x.number]:
+                    dominated_count[y] -= 1
+
+            assert non_dominated_population
+            population_per_rank.append(non_dominated_population)
+
+        return population_per_rank
+
+    def after_trial(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        state: TrialState,
+        values: Optional[Sequence[float]],
+    ) -> None:
+        assert state in [TrialState.COMPLETE, TrialState.FAIL, TrialState.PRUNED]
+        if self._constraints_func is not None:
+            _process_constraints_after_trial(self._constraints_func, study, trial, state)
+        self._random_sampler.after_trial(study, trial, state, values)
+
+
+def multi_choose(n: int, k: int) -> int:
+    return math.factorial(n + k - 1) // math.factorial(k) // math.factorial(n - 1)
+
+
+def generate_default_reference_point(
```

Comment:
It seems that this function is only used in the unit tests. Do you intend to call this function in the `__init__` of the `NSGAIIISampler`?

Reply:
Users can use arbitrary point set as a reference point set according to their preference and give them `NSGAIIISampler`. Since `generate_default_reference_point` only provides reference point sets in default setting, I thought it will be easier to separate from `__init__`. Is it better to move?